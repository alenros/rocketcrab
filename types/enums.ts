export enum PartyStatus {
    loading = "loading",
    party = "party",
    ingame = "ingame",
}

export enum GameStatus {
    loading = "loading",
    inprogress = "inprogress",
    waitingforhost = "waitingforhost",
    error = "error",
}

export enum SocketEvent {
    CONNECT = "connect",
    DISCONNECT = "disconnect",
    RECONNECT = "reconnect",
    GAME_EXIT = "GAME_EXIT",
    GAME_SELECT = "GAME_SELECT",
    GAME_START = "GAME_START",
    HOST_GAME_LOADED = "HOST_GAME_LOADED",
    INVALID_NAME = "INVALID_NAME",
    INVALID_PARTY = "INVALID_PARTY",
    JOIN_PARTY = "JOIN_PARTY",
    NAME = "NAME",
    UPDATE = "UPDATE",
    FINDER_SUBSCRIBE = "FINDER_SUBSCRIBE",
    FINDER_UPDATE = "FINDER_UPDATE",
    CHAT_MESSAGE = "CHAT_MESSAGE",
    KICK_PLAYER = "KICK_PLAYER",
    SET_IS_PUBLIC = "SET_IS_PUBLIC",
}

export enum RocketcrabMode {
    MAIN, // will be shown on rocketcrab.com
    KIDS, // will be shown on kids.rocketcrab.com
    ALL, //  do not use. for testing only.
}
