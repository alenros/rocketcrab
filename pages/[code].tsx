import { useCallback, useEffect, useState } from "react";
import { GetServerSidePropsContext, GetServerSideProps } from "next";
import { useRouter } from "next/router";

import PageLayout from "../components/layout/PageLayout";
import PartyScreen from "../components/party/PartyScreen";
import NameEntry from "../components/party/NameEntry";
import GameLayout from "../components/layout/GameLayout";

import { ClientGameLibrary, ClientParty } from "../types/types";
import { parseCookies } from "nookies";

import { useRocketcrabClientSocket } from "../utils/useRocketcrabClientSocket";
import { useChat } from "../utils/useChat";
import { RocketcrabMode } from "../types/enums";
import { GAME_LIBRARY } from "../config";

export const Code = ({
    gameLibrary = { gameList: [], categories: [] },
    lastPartyState,
    isReconnect,
}: CodeProps): JSX.Element => {
    const router = useRouter();
    const mode = router.locale as RocketcrabMode;
    const code = router?.query?.code as string;

    const previousName = lastPartyState?.me?.name;

    const {
        partyState,
        onNameEntry,
        onSelectGame,
        onStartGame,
        onExitGame,
        onHostGameLoaded,
        showReconnecting,
        onSendChat,
        onKick,
        onSetIsPublic,
    } = useRocketcrabClientSocket({
        code,
        router,
        cookiePartyState: lastPartyState,
        isReconnect,
        gameLibrary,
    });

    const { status, me, chat } = partyState || {};

    const { unreadMsgCount, clearUnreadMsgCount, newestMsg } = useChat(
        chat,
        me
    );

    const [myLastValidName, setMyLastValidName] = useState("");
    useEffect(() => {
        if (me?.name) {
            setMyLastValidName(me?.name);
        } else if (previousName) {
            setMyLastValidName(previousName);
        }
    }, [me?.name, previousName]);

    const [deemphasize, setDeemphasize] = useState(false);
    const onInOutParty = useCallback(
        (outOfParty) => setDeemphasize(outOfParty),
        [setDeemphasize]
    );

    const showLoading = status === "loading";
    const showNameEntry = !showLoading && !me?.name;
    //const showParty = !showLoading && !showNameEntry && status === "party";
    const showGame = !showLoading && !showNameEntry && status === "ingame";

    if (showGame) {
        return (
            <GameLayout
                partyState={partyState}
                onExitGame={onExitGame}
                onStartGame={onStartGame}
                onHostGameLoaded={onHostGameLoaded}
                onSendChat={onSendChat}
                gameLibrary={gameLibrary}
                thisPlayer={me}
                reconnecting={showReconnecting}
                onKick={onKick}
                unreadMsgCount={unreadMsgCount}
                newestMsg={newestMsg}
                clearUnreadMsgCount={clearUnreadMsgCount}
                mode={mode}
            />
        );
    }

    return (
        <PageLayout
            path={code as string}
            loading={showLoading}
            deemphasize={deemphasize}
            reconnecting={showReconnecting}
            mode={mode}
        >
            {showNameEntry ? (
                <NameEntry
                    onNameEntry={onNameEntry}
                    previousName={myLastValidName}
                />
            ) : (
                <PartyScreen
                    partyState={partyState}
                    thisPlayer={me}
                    gameLibrary={gameLibrary}
                    onSelectGame={onSelectGame}
                    onStartGame={onStartGame}
                    resetName={() => onNameEntry("")}
                    onInOutParty={onInOutParty}
                    onSendChat={onSendChat}
                    onKick={onKick}
                    unreadMsgCount={unreadMsgCount}
                    clearUnreadMsgCount={clearUnreadMsgCount}
                    onSetIsPublic={onSetIsPublic}
                />
            )}
        </PageLayout>
    );
};

export const getServerSideProps: GetServerSideProps = async (
    ctx: GetServerSidePropsContext
): Promise<any> => {
    const code = ctx.query?.code;
    const mode = ctx.locale as RocketcrabMode;

    let lastPartyState: ClientParty = { me: {} } as ClientParty;
    let isReconnect = false;

    try {
        lastPartyState = JSON.parse(
            parseCookies(ctx).lastPartyState
        ) as ClientParty;

        isReconnect = lastPartyState.code === code;
        if (!isReconnect) {
            lastPartyState.me.id = null;
        }
        // eslint-disable-next-line no-empty
    } catch (error) {}

    return {
        props: {
            gameLibrary: GAME_LIBRARY[mode],
            ...(lastPartyState ? { lastPartyState } : {}),
            isReconnect,
        },
    };
};

type CodeProps = {
    gameLibrary: ClientGameLibrary;
    lastPartyState?: ClientParty;
    previousId?: string;
    isReconnect: boolean;
};

export default Code;
