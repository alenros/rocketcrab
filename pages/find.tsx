import PrimaryButton from "../components/common/PrimaryButton";
import ButtonGroup from "../components/common/ButtonGroup";
import PageLayout from "../components/layout/PageLayout";
import React, { useEffect, useState } from "react";
import { postJson } from "../utils/utils";
import { useRouter } from "next/router";
import { io } from "socket.io-client";
import { RocketcrabMode, SocketEvent } from "../types/enums";
import {
    ClientGameLibrary,
    FinderState,
    FINDER_ACTIVE_MS,
} from "../types/types";
import { GetServerSideProps } from "next";
import PublicGame from "../components/find/PublicGame";
import Swal from "sweetalert2";
import GameDetail from "../components/detail/GameDetail";
import { FinderInfoCard } from "../components/find/FinderInfoCard";
import { GAME_LIBRARY } from "../config";

const socket = io();

export const Find = ({
    gameLibrary = { gameList: [], categories: [] },
    mode,
}: FindProps): JSX.Element => {
    const router = useRouter();
    const [newLoading, setNewLoading] = useState(false);
    const [showReconnecting, setShowReconnecting] = useState(false);
    const [finderState, setFinderState] = useState<FinderState | undefined>();
    const [gameInfoVisible, setGameInfoVisible] = useState("");

    const { isActive, publicPartyList, finderActiveDates, subscriberCount } =
        finderState ?? {};

    const onClickNew = async (e) => {
        e.preventDefault();
        setNewLoading(true);

        const result = await postJson("/api/new-public").catch(() => {
            setNewLoading(false);
            Swal.fire({
                title: "Try again",
                text:
                    "The server is not allowing public parties to be created right now... maybe you were just a smidge too early? ðŸ˜Š",
                icon: "error",
                heightAuto: false,
            });
        });

        if (result) {
            const { code } = result;

            router.push("/" + code);
        }
    };

    useEffect(() => {
        socket.open();

        socket.emit(SocketEvent.FINDER_SUBSCRIBE);

        socket.on(SocketEvent.FINDER_UPDATE, (newFinderState: FinderState) => {
            setFinderState(newFinderState);
            setShowReconnecting(false);
        });

        socket.on(SocketEvent.DISCONNECT, (reason) => {
            // if the disconnection was initiated by the server
            if (reason === "io server disconnect") {
                // reconnect manually
                socket.connect();
            }
            // else the socket will automatically try to reconnect

            setShowReconnecting(true);
        });

        return () => {
            socket.close();
            setFinderState(undefined);
        };
    }, []);

    if (gameInfoVisible) {
        const selectedGame = gameLibrary.gameList.find(
            ({ id }) => id === gameInfoVisible
        );
        return (
            <PageLayout
                reconnecting={showReconnecting}
                deemphasize={true}
                mode={mode}
            >
                <div style={{ textAlign: "center" }}>
                    <GameDetail
                        game={selectedGame}
                        allCategories={gameLibrary.categories}
                    />
                    <PrimaryButton onClick={() => setGameInfoVisible("")}>
                        Back to Public Parties
                    </PrimaryButton>
                </div>
            </PageLayout>
        );
    }

    return (
        <PageLayout reconnecting={showReconnecting} mode={mode}>
            <div className="description">Public Parties (beta)</div>
            {isActive ? (
                <>
                    {publicPartyList?.length ? (
                        publicPartyList?.map((party) => (
                            <PublicGame
                                key={party.code}
                                party={party}
                                gameLibrary={gameLibrary}
                                onWhatIs={setGameInfoVisible}
                            />
                        ))
                    ) : (
                        <div style={{ textAlign: "center" }}>
                            No public parties found. ðŸ˜ž{" "}
                            <div style={{ display: "inline-block" }}>
                                You should make one! ðŸ¥°
                            </div>
                        </div>
                    )}
                    <FinderInfoCard
                        subscriberCount={subscriberCount}
                        subscriberCountMsg="here still looking for a party to choose!"
                        showCountdown={!!finderActiveDates?.lastStart}
                        countdownStart={
                            finderActiveDates.lastStart + FINDER_ACTIVE_MS
                        }
                        countdownMsg="Public parties will close"
                    />
                </>
            ) : (
                <div style={{ textAlign: "center" }}>
                    <div>
                        To make sure there are enough players, public games open
                        roughly every four hours from Thursday to Saturday.
                    </div>
                    <FinderInfoCard
                        subscriberCount={subscriberCount}
                        subscriberCountMsg="waiting on this page."
                        showCountdown={!!finderActiveDates?.nextStart}
                        countdownStart={finderActiveDates?.nextStart}
                        countdownMsg="Public parties will open next"
                        findTimeDates={finderActiveDates?.nextWeekOfStarts}
                    />
                </div>
            )}

            <ButtonGroup>
                <PrimaryButton href="/" size="large">
                    Back
                </PrimaryButton>

                {isActive && (
                    <PrimaryButton
                        onClick={onClickNew}
                        loading={newLoading}
                        size="large"
                    >
                        Start Public Party
                    </PrimaryButton>
                )}
            </ButtonGroup>
            <style jsx>{`
                .description {
                    font-size: 1.1em;
                    text-align: center;
                    margin-bottom: 1em;
                }
            `}</style>
        </PageLayout>
    );
};

export const getStaticProps: GetServerSideProps = async ({ locale }) => ({
    props: {
        gameLibrary: GAME_LIBRARY[locale],
        mode: locale as RocketcrabMode,
    },
});

type FindProps = {
    gameLibrary: ClientGameLibrary;
    mode: RocketcrabMode;
};

export default Find;
