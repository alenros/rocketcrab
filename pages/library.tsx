import { GetServerSideProps } from "next";
import { useRouter } from "next/router";
import React, { useCallback } from "react";
import GameLibrary from "../components/library/GameLibrary";
import PageLayout from "../components/layout/PageLayout";
import { getClientGameLibrary } from "../config";
import { ClientGameLibrary } from "../types/types";
import { useLibraryState } from "../utils/utils";
import { RocketcrabMode } from "../types/enums";

const CLIENT_GAME_LIBRARY = getClientGameLibrary(RocketcrabMode.MAIN);

export const Library = ({ gameLibrary }: LibraryProps): JSX.Element => {
    const router = useRouter();

    const onDone = useCallback(() => router.back(), [router]);

    const onViewGame = useCallback(
        (gameId) => {
            router.push("/game/[gameid]", "/game/" + gameId);
        },
        [router]
    );

    const libraryState = useLibraryState();

    return (
        <PageLayout>
            <GameLibrary
                gameLibrary={gameLibrary}
                backToLabel="rocketcrab.com"
                onDone={onDone}
                setViewingGameId={onViewGame}
                libraryState={libraryState}
            />
        </PageLayout>
    );
};

type LibraryProps = {
    gameLibrary: ClientGameLibrary;
};

export const getStaticProps: GetServerSideProps = async () => ({
    props: {
        gameLibrary: CLIENT_GAME_LIBRARY,
    },
});

export default Library;
