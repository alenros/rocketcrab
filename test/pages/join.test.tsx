import React from "react";
import { render } from "../testUtils";
import { Join } from "../../pages/join";
import { RocketcrabMode } from "../../types/enums";

jest.mock("next/router", () => ({
    useRouter: () => ({ query: { invalid: "wxyz" }, locale: "MAIN" }),
}));

describe("pages/join.tsx", () => {
    it("matches snapshot", () => {
        const { asFragment } = render(<Join mode={RocketcrabMode.MAIN} />, {});
        expect(asFragment()).toMatchSnapshot();
    });
});
