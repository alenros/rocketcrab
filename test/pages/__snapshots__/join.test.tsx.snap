// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/join.tsx matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-911348231 container"
  >
    <main
      class="jsx-911348231"
    >
      <div
        class="jsx-911348231 main-content"
      >
        <div
          class="jsx-1758552329 title"
        >
          <div
            class="jsx-1758552329"
            style="margin: .5em;"
          >
            <img
              class="jsx-1758552329 rocket"
              src="/rocket.svg"
            />
            <img
              class="jsx-1758552329 crab"
              src="/crab.svg"
            />
          </div>
          <div>
            <div
              style="display: inline-block; white-space: nowrap;"
            >
              <div
                class="jsx-418573366 tooltip "
              >
                <h2
                  class="jsx-1758552329 party-url"
                >
                  rocketcrab.com
                </h2>
              </div>
            </div>
          </div>
        </div>
        <div
          class="jsx-2322376654 description"
        >
          Join Party
        </div>
        <div
          class="jsx-2322376654 input-container"
        >
          <div
            class="jsx-1274958171 with-label"
          >
            <div
              class="jsx-1274958171 input-container "
            >
              <div
                class="jsx-1274958171 input-wrapper hover  "
              >
                <input
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="jsx-1274958171  "
                  maxlength="4"
                  placeholder="ex. abcd"
                  type="text"
                  value=""
                />
                <div
                  class="jsx-1724197483 clear-icon "
                >
                  <svg
                    class="jsx-1724197483"
                    fill="none"
                    shape-rendering="geometricPrecision"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="1.5"
                    viewBox="0 0 24 24"
                  >
                    <path
                      class="jsx-1724197483"
                      d="M18 6L6 18"
                    />
                    <path
                      class="jsx-1724197483"
                      d="M6 6l12 12"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="jsx-2322376654 description"
          style="color: red;"
        >
          wxyz does not exist üòû
        </div>
        <div
          class="jsx-4094718877 item justify jsx-1720607024 "
        >
          <div
            class="jsx-2819926550 item  jsx-4216129763 "
          >
            <div>
              <button
                class="jsx-2230649419 btn "
                href="/"
                style="width: auto;"
                type="button"
              >
                <div
                  class="text"
                >
                  Back
                </div>
              </button>
            </div>
          </div>
          <div
            class="jsx-2819926550 item  jsx-4216129763 "
          >
            <button
              class="jsx-512623370 btn "
              disabled=""
              style="width: auto;"
              type="button"
            >
              <div
                class="text"
              >
                Join
              </div>
            </button>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="jsx-3899203255"
      style="margin-top: 1em;"
    >
      <div
        class="jsx-3899203255 side"
      />
      <div
        class="jsx-3899203255 tag"
      >
        rocketcrab by 
        <a
          class="jsx-3899203255"
          href="https://www.tannerkrewson.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Tanner Krewson
        </a>
        <br
          class="jsx-3899203255"
        />
        <a
          class="jsx-3899203255"
          href="https://github.com/tannerkrewson/rocketcrab"
          rel="noopener noreferrer"
          target="_blank"
        >
          View on GitHub
        </a>
      </div>
      <div
        class="jsx-3899203255 side"
      >
        <div
          class="jsx-3899203255 theme-toggle"
        >
          <div
            class="react-toggle react-toggle--checked"
          >
            <div
              class="react-toggle-track"
            >
              <div
                class="react-toggle-track-check"
              >
                <span
                  class="jsx-791876786 sun-moon"
                >
                  ‚òÄÔ∏è
                </span>
              </div>
              <div
                class="react-toggle-track-x"
              >
                <span
                  class="jsx-791876786 sun-moon"
                >
                  üåô
                </span>
              </div>
            </div>
            <div
              class="react-toggle-thumb"
            />
            <input
              checked=""
              class="react-toggle-screenreader-only"
              type="checkbox"
            />
          </div>
        </div>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
